libkbfs/folder_branch_ops.go

```diff
@@ -1782,34 +1782,38 @@ func (fbo *folderBranchOps) isRevisionConflict(err error) bool {
 }
 
 func (fbo *folderBranchOps) finalizeMDWriteLocked(ctx context.Context,
-	lState *lockState, md *RootMetadata, bps *blockPutState) (err error) {
+	lState *lockState, md *RootMetadata, bps *blockPutState, excl EXCL) (err error) {
 	fbo.mdWriterLock.AssertLocked(lState)
 
 	// finally, write out the new metadata
 	mdops := fbo.config.MDOps()
 
-	doUnmergedPut, wasMasterBranch := true, fbo.isMasterBranchLocked(lState)
+	doUnmergedPut, wasMasterBranch := (excl == NoEXCL), fbo.isMasterBranchLocked(lState)
 	mergedRev := MetadataRevisionUninitialized
 
 	if fbo.isMasterBranchLocked(lState) {
 		// only do a normal Put if we're not already staged.
 		err = mdops.Put(ctx, md)
-		doUnmergedPut = fbo.isRevisionConflict(err)
-		if err != nil && !doUnmergedPut {
-			return err
-		}
-		// The first time we transition, our last known MD revision is
-		// the same (at least) as what we thought our new revision
-		// should be.  Otherwise, just leave it at uninitialized and
-		// let the resolver sort it out.
-		if doUnmergedPut {
+
+		if doUnmergedPut = fbo.isRevisionConflict(err); doUnmergedPut {
 			fbo.log.CDebugf(ctx, "Conflict: %v", err)
 			mergedRev = md.Revision
+
+			if excl == WithEXCL {
+				// If this was caused by an exclusive create, we shouldn't do an
+				// UnmergedPut, but rather wait until we are on master branch.
+				return err
+			}
+		} else if err != nil {
+			return err
 		}
+	} else if excl == WithEXCL {
+		return UnmergedError{}
```

Comment:
A more specific error might be nice here.  Either a new `UnmergedEXCLError`, or you can just use `errors.New()` since this is a pretty unexpected case and I doubt we'll ever need to check for the error type.